File: brian2/codegen/languages/cpp/templates/synapses.cpp
////////////////////////////////////////////////////////////////////////////
//// MAIN CODE /////////////////////////////////////////////////////////////

{% macro main() %}
	// USE_SPECIFIERS { _spiking_synapses, _postsynaptic,_presynaptic }

    //// SUPPORT CODE //////////////////////////////////////////////////////////
	{% for line in support_code_lines %}
	// {{line}}
	{% endfor %}

	////// HANDLE DENORMALS ///
	{% for line in denormals_code_lines %}
	{{line}}
	{% endfor %}

	////// HASH DEFINES ///////
	{% for line in hashdefine_lines %}
	{{line}}
	{% endfor %}

	///// POINTERS ////////////
	{% for line in pointers_lines %}
	{{line}}
	{% endfor %}

	//// MAIN CODE ////////////
	for(int _spiking_synapse_idx=0;
		_spiking_synapse_idx<_num_spiking_synapses;
		_spiking_synapse_idx++)
	{
		const int _neuron_idx = _spiking_synapses[_spiking_synapse_idx];
		const int _postsynaptic_idx = _postsynaptic[_neuron_idx];
		const int _presynaptic_idx = _presynaptic[_neuron_idx];
		{% for line in code_lines %}
		{{line}}
		{% endfor %}
	}
{% endmacro %}

////////////////////////////////////////////////////////////////////////////
//// SUPPORT CODE //////////////////////////////////////////////////////////

{% macro support_code() %}
	{% for line in support_code_lines %}
	{{line}}
	{% endfor %}
{% endmacro %}

File: brian2/codegen/languages/python/templates/synapses.py
# USE_SPECIFIERS { _synaptic_post, _spiking_synapses }

import numpy as np

_post_neurons = _synaptic_post.take(_spiking_synapses)
_perm = _post_neurons.argsort()
_aux = _post_neurons.take(_perm)
_flag = np.empty(len(_aux)+1, dtype=bool)
_flag[0] = _flag[-1] = 1
np.not_equal(_aux[1:], _aux[:-1], _flag[1:-1])
_F = _flag.nonzero()[0][:-1]
np.logical_not(_flag, _flag)
while len(_F):
    _u = _aux.take(_F)
    _i = _perm.take(_F)
    _postsynaptic_idx = _u
    _neuron_idx = _spiking_synapses[_i]
    # TODO: how do we get presynaptic indices? do we need to?

    {% for line in code_lines %}
    {{line}}
    {% endfor %}

    _F += 1
    _F = np.extract(_flag.take(_F), _F)
